platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongList.java

```diff
@@ -291,26 +390,109 @@ protected final void writeHeader(final FileChannel fc) throws IOException {
+            logger.warn("New min valid index cannot exceed current list size, returning.");
+            return;
+        }
+
+        // This is an optimization: we know that a chunk for this index must exist,
+        // so let's make sure that it's there
+        createOrGetChunk(newMinValidIndex);
+        shrinkIfNeeded(newMinValidIndex);
+        minValidIndex.set(newMinValidIndex);
+    }
+
+    /**
+     * Expand the available data storage if needed to allow storage of an item at newIndex
+     *
+     * @param newIndex the index of the new item we would like to add to storage
+     */
+    protected C createOrGetChunk(final long newIndex) {
+        size.getAndUpdate(oldSize -> newIndex >= oldSize ? (newIndex + 1) : oldSize);
+        final int chunkIndex = (int) (newIndex / numLongsPerChunk);
+        final C result = chunkList.get(chunkIndex);
+        if (result == null) {
+            final C newChunk = createChunk();
+            // set new chunk if it's not created yet, if it is - release the chunk immediately
+            // and use the one from the list
+            final C oldChunk = chunkList.compareAndExchange(chunkIndex, null, newChunk);
+            if (oldChunk == null) {
+                return newChunk;
+            } else {
+                fullChunkCleanup(newChunk);
+                return oldChunk;
+            }
+        } else {
+            return result;
+        }
+    }
+
+    /**
+     * Deletes values up to {@code newMinValidIndex}, releases memory chunks reserved for these
+     * values. Note that it takes {@code reservedBufferOffset} into account to decide if a chunk has
+     * to be released.
+     *
+     * @param newMinValidIndex new minimal valid index, left boundary of the list
+     */
+    private void shrinkIfNeeded(final long newMinValidIndex) {
+        final int firstValidChunkWithBuffer =
+                (int) Math.max((newMinValidIndex - reservedBufferLength) / numLongsPerChunk, 0);
+        final int firstChunkIndexToDelete = firstValidChunkWithBuffer - 1;
+        for (int i = firstChunkIndexToDelete; i >= 0; i--) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                fullChunkCleanup(chunk);
+            }
+        }
+
+        // clean up a chunk with data
+        final int firstChunkWithData = (int) (newMinValidIndex / numLongsPerChunk);
+        final long numberOfElementsToCleanUp = (newMinValidIndex % numLongsPerChunk);
+        partialChunkCleanup(firstChunkWithData, numberOfElementsToCleanUp);
+
+        // clean up chunk(s) reserved for buffer
+        for (int i = firstValidChunkWithBuffer; i < firstChunkWithData; i++) {
+            partialChunkCleanup(i, numLongsPerChunk);
+        }
+    }
+
+    protected abstract void fullChunkCleanup(C chunk);
+
+    protected abstract void partialChunkCleanup(final int chunkIndex, final long entriesToCleanUp);
```

Comment:
Can these two methods be combined together? Full cleanup is a partial cleanup with entriesToCleanup set to numLongsPerChunk, isn't it?

Reply:
Not sure what's the benefit of combining these two methods. 
They serve slightly different purposes:
 - `fullChunkCleanup` releases the chunk and let the memory (be it RAM or disk) to be reclaimed
 - `partialChunkCleanup` just zeroes some values in the chunk to follow the `minValidIndex` contract